### ISSUE TITLE
Function marked as irrelevant when it isn't

### ISSUE THREAD

**Issue by gallais** [agda/agda#3200]
This is a reduced version of agda/agda-stdlib/issues/429. All my attempts to simplify
it even more make the problem go away.

It is a regression in 2.6.0: the code works just fine in 2.5.4.1 (which is what we use
on travis to check the stdlib).

The `test` fails with the error (**NB:** note the dot before `just` and the fact the redex
`From-just (just _)` doesn't fire) :

```
From-just
.((just
   (λ b →
      subst (_≡ true)
      (sym (proj₂ (ID (λ b₁ → if b₁ then true else true)) b))
      ((λ { true → subst (λ b₁ → b₁ ≡ true) (sym refl) refl
          ; false → subst (λ b₁ → b₁ ≡ true) (sym refl) refl
          })
       b))))
!=< (b : Bool) → (if b then true else true) ≡ true of type Set
when checking that the expression
prove (λ b → if b then true else true) has type
(b : Bool) → (if b then true else true) ≡ true
```

At some point I had a more complex version where I couldn't even define `prove`
because trying to give `from-just` the expression `solve2 f` would yield an error
telling me `solve2` was defined irrelevant but used in a relevant context.

```agda
open import Data.Bool
open import Data.Maybe.Base
open import Relation.Binary.PropositionalEquality
open import Function
open import Data.Product

ID : (f : Bool → Bool) → Σ[ g ∈ (Bool → Bool) ] ∀ b → f b ≡ g b
ID f = (λ b → f b) , λ b → refl

solve1 : (f : Bool → Bool) → Maybe (∀ b → f b ≡ true)
solve1 f with f true | inspect f true | f false | inspect f false
... | true | p | true | q = just let open Reveal_·_is_ in λ where
  true  → subst (λ b → b ≡  true) (sym (eq p)) refl
  false → subst (λ b → b ≡  true) (sym (eq q)) refl
... | _ | _ | _ | _ = nothing

solve2 : (f : Bool → Bool) → Maybe (∀ b → f b ≡ true)
solve2 f =
  let (g , eq) = ID f in
  case solve1 g of λ where
    (just hyp) → just λ b → subst (_≡  true) (sym (eq b)) (hyp b)
    nothing    → nothing

prove : ∀ f → From-just (solve2 f)
prove f = from-just (solve2 f)

test : ∀ b → (if b then true else true) ≡ true
test = prove (λ b → if b then true else true)
```

Note that if we use `with` instead of `let` + `case_of_` to define `solve2` then
everything is fine:

```
solve2 : (f : Bool → Bool) → Maybe (∀ b → f b ≡ true)
solve2 f with ID f
... | (g , eq) with solve1 g
solve2 f | g , eq | just hyp = just λ b → subst (_≡  true) (sym (eq b)) (hyp b)
solve2 f | g , eq | nothing = nothing
```

**Comment by gallais** [comment]
Reduced, self-contained test-case.

Note that if you drop `let record {} = tt in` then it goes through just fine!

```agda
open import Agda.Builtin.Bool
open import Agda.Builtin.Unit

data Maybe (A : Set) : Set where
  just : A → Maybe A
  nothing : Maybe A

case_of_ : ∀ {a b} {A : Set a} {B : Set b} → A → (A → B) → B
case x of f = f x

solve2 : Maybe (Bool → Bool)
solve2 =
  let record {} = tt in
  case true of λ where
    true → just (λ x → true)
    false → nothing

module _ {A : Set} where

  From-just : Maybe A → Set
  From-just (just _) = A
  From-just nothing  = ⊤

  from-just : (x : Maybe A) → From-just x
  from-just (just x) = x
  from-just nothing  = _

test : Bool
test = from-just solve2 true
```

**Comment by HuStmpHrrr** [comment]
this version also gives the same error
```
Agda version 2.6.0-67f316e
```

let me try even earlier one.

**Comment by gallais** [comment]
@HuStmpHrrr I'm currently running a git bisect to pinpoint the commit where it started to fail.

There are less than 7 steps left so it should converge in ~1h30m maximum.

**Comment by gallais** [comment]
git bisect points at @andreasabel's a1030da59d1726b5b4ee6c82028434e05a6d584b

**Comment by andreasabel** [comment]
Some observations:
* The blamed commit removed the abuse of `Prop` for dummy types.  However, dummy types still have sort `Prop`. Together with @jespercockx resurrection of  `Prop` it likely triggered this issue.
* What is happening is that after checking pattern `record {}` the `envRelevance` is `Irrelevant`.  This does not look right.  
* Records are treated as irrelevant if all their components are irrelevant (which is trivially true for empty records).  However, this should not matter.  Indeed, the test still fails if we use a non-empty record:
```agda
record R : Set where
  field
    fld : Bool
...
  let record { fld = bla } = record { fld = true } in
...
```
* With `--no-prop` (and respecting this flag in LHS checking), the issue goes away.

**Comment by andreasabel** [comment]
I will fix this by completing the #3107 eradication of the `Sort Prop` hack.

**Comment by gallais** [comment]
Thanks!
